src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/RazorCodeLensProvider.ts

```diff
@@ -0,0 +1,130 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+import * as vscode from 'vscode';
+import { isRazorCSharpFile, getRazorDocumentUri } from './RazorConventions';
+import { RazorLanguageFeatureBase } from './RazorLanguageFeatureBase';
+import { LanguageKind } from './RPC/LanguageKind';
+import { CancellationToken } from 'vscode-jsonrpc';
+
+class RazorCodeLens extends vscode.CodeLens {
+    constructor(
+        range: vscode.Range,
+        public uri: vscode.Uri,
+        public document: vscode.TextDocument,
+        command?: vscode.Command) {
+
+        super(range, command);
+    }
+}
+
+export class RazorCodeLensProvider
+    extends RazorLanguageFeatureBase
+    implements vscode.CodeLensProvider {
+
+    public async provideCodeLenses(document: vscode.TextDocument, token: vscode.CancellationToken) {
+        try {
+            const razorDocument = await this.documentManager.getDocument(document.uri);
+            if (!razorDocument) {
+                return;
+            }
+
+            const csharpDocument = razorDocument.csharpDocument;
+
+            // Get all the code lenses that applies to our projected C# document.
+            const codeLenses = await vscode.commands.executeCommand<vscode.CodeLens[]>(
+                'vscode.executeCodeLensProvider',
+                csharpDocument.uri) as vscode.CodeLens[];
+
+            // Re-map the CodeLens locations to the original Razor document.
+            const remappedCodeLenses = new Array<vscode.CodeLens>();
+            for (const codeLens of codeLenses) {
+                const result = await this.serviceClient.mapToDocumentRange(
+                    LanguageKind.CSharp,
+                    codeLens.range,
+                    razorDocument.uri);
+                if (result) {
+                    const newCodeLens = new RazorCodeLens(result.range, razorDocument.uri, document, codeLens.command);
+                    remappedCodeLenses.push(newCodeLens);
+                } else {
+                    // This means this CodeLens was for non-user code. We can safely ignore those.
+                }
+            }
+
+            return remappedCodeLenses;
+
+        } catch (error) {
+            return [];
+        }
+    }
+
+    public async resolveCodeLens(codeLens: vscode.CodeLens, token: vscode.CancellationToken) {
+        if (codeLens instanceof RazorCodeLens) {
+            return await this.resolveRazorCodeLens(codeLens, token);
+        }
+    }
+
+    private async resolveRazorCodeLens(codeLens: RazorCodeLens, token: CancellationToken): Promise<vscode.CodeLens> {
+        // Initialize with default values.
```

Comment:
Hmm curious, why do we throw away the command that's associated?

Reply:
At this point, the codelens.command will be undefined. Nothing is thrown away.

Reply:
Oh interesting. What does the command in provide CodeLens above get used for then? 

I imagined that provide CodeLens was the piece that populated the UI and resolve responded to clicking the UI my b!

Reply:
provideCodeLens doesn't really set anything in command. I initially assumed it was for populating the UI but it is just for VSCode to know where to populate. When we scroll down to that area, we get a call to resolveCodeLens which is why you see the exact number of references without even clicking on it.